src/lib.rs

```diff
@@ -822,10 +822,44 @@ pub mod sync {
         /// cell.set("hello".to_string()).unwrap();
         /// assert_eq!(cell.into_inner(), Some("hello".to_string()));
         /// ```
-        pub fn into_inner(self) -> Option<T> {
-            // Because `into_inner` takes `self` by value, the compiler statically verifies
-            // that it is not currently borrowed. So it is safe to move out `Option<T>`.
-            self.0.value.into_inner()
+        pub fn into_inner(mut self) -> Option<T> {
+            // Because `into_inner` takes `self` by value, the compiler
+            // statically verifies that it is not currently borrowed.
+            // So it is safe to move out `Option<T>`.
+            let value = self.take_inner();
+
+            // We just moved out the value without marking this `OnceCell` as
+            // uninitialized, that's why we should mark it as such now to avoid
+            // dropping uninitialized memory in the `Drop` implementation
+            // and the end of this function.
+            // The reason we're doing this here, but not in `take_inner` is
+            // to avoid unnecessary performance penalties on usual `Drop` calls.
+            // Safe b/c we have a unique access.
+            unsafe { self.0.mark_uninitialized() };
+
+            value
+        }
+
+        /// Takes the wrapped value out of a `OnceCell`.
+        /// Afterwards the cell is no longer initialized, and must be free'd
+        /// WITHOUT dropping. Only used by `into_inner` and `drop`.
+        fn take_inner(&mut self) -> Option<T> {
```

Comment:
This function takes out the value of the cell without changing initialized flag, this could lead to double-free even in simple code like this:
```rust
let cell = OnceCell::new();
cell.set(vec![0]);
let inner = cell.take_inner();
// drop inner
// drop cell => drop inner again
```

Despite the fact that this is private API, this function must be `unsafe`.